telemetry/writer.go

```diff
@@ -45,12 +45,19 @@ func (t *writer) WriteStatus(ev *progrock.StatusUpdate) error {
 
 	for _, m := range ev.Memberships {
 		for _, vid := range m.Vertexes {
-			t.maybeEmitOp(ts, vid, false)
+			if v, found := t.pipeliner.Vertex(vid); found {
+				t.maybeEmitOp(ts, v, false)
+			}
 		}
 	}
-
-	for _, v := range ev.Vertexes {
-		t.maybeEmitOp(ts, v.Id, true)
+	for _, eventVertex := range ev.Vertexes {
+		if v, found := t.pipeliner.Vertex(eventVertex.Id); found {
+			// override the vertex with the current event vertex since a
+			// single PipelineEvent could contain duplicated vertices with
+			// different data like started and completed
+			v.Vertex = eventVertex
+			t.maybeEmitOp(ts, v, true)
```

Comment:
Is the idea here to emit every intermediate state for the vertex rather than the final one after the pipeliner handles the update?

Reply:
we need to emit all the states (start, completed) for each vertex so that cloud can aggregate the data correctly. The specific bug that we're currently hitting is a single progrock update event containing:
```
vertices: [
{"id": "sha256:1", "started": 4567, "completed: 891011},
{"id": "sha256:1", "started": 1234, "completed: 5678},
]
```
Given the current logic, `pipeliner` will only cache the first vertex it's seeing (started 4567) so when this method (https://github.com/dagger/dagger/pull/5573/files#diff-e9441eace2cdf8e0d9b2de26f0a594c74317c27faad4a847ff6555b98ff26a62L53) is called, `pipeliner.Vertex(id)` always return the same (incorrect) vertex. 

With this change we're sure that all vertices that we get in the progrock event are emitted.